ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -0,0 +1,351 @@
+    fprintf(stderr, "sampling_overshoot_time=%ld\n", sampling_overshoot_time_ns);
+    fprintf(stderr, "working_window_time=%ld\n", working_window_time_ns);
+    fprintf(stderr, "sampling_interval=%zu\n", sampler->sampling_interval);
+    fprintf(stderr, "sampling_probability=%f\n", sampler->sampling_probability);
+    fprintf(stderr, "expected allocs in 60s=%f\n", allocs_in_60s);
+    fprintf(stderr, "expected samples in 60s=%f\n", samples_in_60s);
+    fprintf(stderr, "expected sampling time in 60s=%f (previous real=%f)\n", expected_total_sampling_time_in_60s, real_total_sampling_time_in_60s);
+    fprintf(stderr, "target_overhead=%f\n", sampler->target_overhead);
+    fprintf(stderr, "target_overhead_adjustment=%f\n", sampler->target_overhead_adjustment);
+    fprintf(stderr, "target_sampling_time=%ld\n", target_sampling_time_ns);
+    fprintf(stderr, "expected max overhead in 60s=%f\n", target_overhead / 100.0 * 60);
+    fprintf(stderr, "-------\n");
+  #endif
+
+  sampler->events_since_last_readjustment = 0;
+  sampler->samples_since_last_readjustment = 0;
+  sampler->sampling_time_since_last_readjustment_ns = 0;
+  sampler->last_readjust_time_ns = now;
+  sampler->has_completed_full_adjustment_window = true;
+}
+
+// ---
+// Below here is boilerplate to expose the above code to Ruby so that we can test it with RSpec as usual.
+
+static VALUE _native_new(VALUE klass);
+static void _native_free(void *state);
+static VALUE _native_reset(VALUE self, VALUE now);
+static VALUE _native_set_overhead_target_percentage(VALUE self, VALUE target_overhead, VALUE now);
+static VALUE _native_should_sample(VALUE self, VALUE now);
+static VALUE _native_after_sample(VALUE self, VALUE now);
+static VALUE _native_probability(VALUE self);
+
+typedef struct sampler_state {
+  discrete_dynamic_sampler sampler;
+} sampler_state;
+
+void collectors_discrete_dynamic_sampler_init(VALUE profiling_module) {
+  VALUE collectors_module = rb_define_module_under(profiling_module, "Collectors");
+  VALUE discrete_sampler_module = rb_define_module_under(collectors_module, "DiscreteDynamicSampler");
+  VALUE testing_module = rb_define_module_under(discrete_sampler_module, "Testing");
+  VALUE sampler_class = rb_define_class_under(testing_module, "Sampler", rb_cObject);
+
+  rb_define_alloc_func(sampler_class, _native_new);
+
+  rb_define_method(sampler_class, "_native_reset", _native_reset, 1);
+  rb_define_method(sampler_class, "_native_set_overhead_target_percentage", _native_set_overhead_target_percentage, 2);
+  rb_define_method(sampler_class, "_native_should_sample", _native_should_sample, 1);
+  rb_define_method(sampler_class, "_native_after_sample", _native_after_sample, 1);
+  rb_define_method(sampler_class, "_native_probability", _native_probability, 0);
+}
+
+static const rb_data_type_t sampler_typed_data = {
+  .wrap_struct_name = "Datadog::Profiling::DiscreteDynamicSampler::Testing::Sampler",
+  .function = {
+    .dfree = _native_free,
```

Comment:
Minor: I'm curious -- an earlier version of the PR used `RUBY_DEFAULT_FREE` here, and it looks to me we could still do that? :thinking:

Reply:
🤦  I think my brain short-circuited yesterday and I thought this was null/noop before.